File: 07 Tree/C#/BinarySearchTree.cs
using System;
using System.Collections.Generic;

namespace Tree
{
    public class BinarySearchTree<T> : ITree<T> where T : IComparable<T>
    {
        private Node _root;

        #region Node inner class
        class Node
        {
            public T data;
            public Node left, right;
        }
        #endregion

        #region public methods

        public void Insert(T val)
        {
            _root = InsertInternal(_root, val);
        }

        public void PrintBFS()
        {
            Queue<Node> childs = new Queue<Node>();
            PrintBFS(_root, childs);
        }

        public void PrintDFSInorder()
        {
            PrintDFSInorder(_root);
        }

        public void BFSWithoutRecursion()
        {
            BFSWithoutRecursionInternal(_root);
        }

        public void Edge()
        {
            InternalPrintLeft(_root);
            LeafNode();
            InternalPrintRight(_root.right);    //skip root as already printed
        }

        public void PrintDFSPostOrder()
        {
            PrintDFSPostOrder(_root);
        }

        public void PrintDFSPreorder()
        {
            PrintDFS(_root);
        }

        public void Remove(T val)
        {
            // find
            Node node = SearchDFS(val);

            // manipulate right if available
            node = RemoveNode(node);
            // or manipulate left if available
            // or just delete
        }

        private Node RemoveNode(Node node)
        {
            if (node.left == null && node.right == null)
            {
                node = null;
                return node;
            }

            if (node.right != null)
            {
                node.data = node.right.data;
                node.right = RemoveNode(node.right);
            }
            else if (node.left != null)
            {
                node.data = node.left.data;
                node.left = RemoveNode(node.left);
            }
            else
            {
                node = null;
            }

            return node;
        }

        private Node SearchDFS(T val)
        {
            return SearchDFSUtil(_root, val);
        }

        private Node SearchDFSUtil(Node root, T val)
        {
            if (root == null)
            {
                return null;
            }

            if (0 == root.data.CompareTo(val))
            {
                return root;
            }

            var temp = SearchDFSUtil(root.left, val);
            if (temp != null)
                return temp;

            var temp1 = SearchDFSUtil(root.right, val);
            if (temp1 != null)
                return temp1;

            return null;
        }

        public T Search()
        {
            throw new NotImplementedException();
        }

        public void LeafNode()
        {
            Queue<Node> childs = new Queue<Node>();
            LeafNodeInternal(_root, childs);
        }

        #endregion

        #region private methods
        private Node InsertInternal(Node current, T val)
        {
            if (current == null)
            {
                current = new Node() { data = val };
            }
            else
            {
                if (1 == current.data.CompareTo(val))
                current.left = InsertInternal(current.left, val);
                else
                    current.right = InsertInternal(current.right, val);
            }
            return current;
        }

        private void PrintBFS(Node root, Queue<Node> childs)
        {
            if (root == null)
                return;

            Console.WriteLine(root.data);
            childs.Enqueue(root.left);
            childs.Enqueue(root.right);

            while (childs.Count > 0)
            {
                PrintBFS(childs.Dequeue(), childs);
            }
        }

        private void PrintDFS(Node root)
        {
            if (root == null)
                return;

            Console.WriteLine(root.data);
            PrintDFS(root.left);
            PrintDFS(root.right);
        }

        private void PrintDFSInorder(Node root)
        {
            if (root == null)
                return;

            PrintDFSInorder(root.left);
            Console.WriteLine(root.data);
            PrintDFSInorder(root.right);
        }

        private void PrintDFSPostOrder(Node root)
        {
            if (root == null)
                return;

            PrintDFSPostOrder(root.left);
            PrintDFSPostOrder(root.right);
            Console.WriteLine(root.data);
        }

        private void LeafNodeInternal(Node root, Queue<Node> childs)
        {
            if (root == null)
                return;
            if (root.left == null && root.right == null)
                Console.WriteLine(root.data);

            childs.Enqueue(root.left);
            childs.Enqueue(root.right);

            while (childs.Count > 0)
            {
                LeafNodeInternal(childs.Dequeue(), childs);
            }
        }

        private void InternalPrintRight(Node root)
        {
            if (root == null)
                return;
            if (root.right == null)  // skip the leaf left
                return;
            InternalPrintRight(root.right); //bottom to top
            Console.WriteLine(root.data);
        }

        private void InternalPrintLeft(Node root)
        {
            if (root == null)
                return;
            if (root.left == null)  // skip the leaf left
                return;
            Console.WriteLine(root.data);
            InternalPrintLeft(root.left);
        }

        private void BFSWithoutRecursionInternal(Node root)
        {
            if (root == null)
                return;
            Queue<Node> nodes = new Queue<Node>();
            Queue<Node> childNodes = new Queue<Node>();
            childNodes.Enqueue(root);

            while (childNodes.Count > 0)
            {
                Node node = childNodes.Dequeue();
                if (node.left != null)
                    childNodes.Enqueue(node.left);
                if (node.right != null)
                    childNodes.Enqueue(node.right);

                nodes.Enqueue(node);
            }
            foreach (var item in nodes)
            {
                Console.WriteLine(item.data);
            }
        }

        #endregion
    }
}

File: 07 Tree/Java/BinarySearchTree.java
package binarySearchTree;

import java.util.ArrayList;
import java.util.List;
import java.util.Stack;

public class BinarySearchTree {

	class Node {
		int data;
		Node left;
		Node right;
		
		@Override
		public String toString() {
			String str= new String(""+data+"");
			return str;
		}
	}

	Node root;

	public void insert(int data) {
		root = insert(data, root);
	}

	public void BFSTraverse() {
		List<Node> childNodes = new ArrayList<>();
		BFStraverse(root, childNodes);
	}

	public void preOrder() {
		preOrder(root);
	}

	public void postOrder() {
		postOrder(root);
	}

	public void inOrder() {
		inOrder(root);
	}

	public void printLeafNodes() {
		List<Node> leafNodes = new ArrayList<>();
		printLeafNodes(root, leafNodes);
	}

	public void printEdgeNodes() {
		printLeftNodes(root);
		printLeafNodes();
		printRightNodes(root.right);
	}
	
	public void BFSWithoutRecursion()
	{
		BFSWithoutRecursion(root);
	}

	
	private Node insert(int data, Node node) {
		if (node == null) {
			node = new Node();
			node.data = data;
			return node;
		} else if (data < node.data) {
			node.left = insert(data, node.left);
		} else if (data > node.data) {
			node.right = insert(data, node.right);
		}
		return node;
	}

	private void preOrder(Node node) {
		if (node == null)
			return;
		System.out.print(node.data + "	");
		preOrder(node.left);
		preOrder(node.right);
	}

	private void postOrder(Node node) {
		if (node == null)
			return;
		postOrder(node.left);
		postOrder(node.right);
		System.out.print(node.data + "	");
	}

	private void inOrder(Node node) {
		if (node == null)
			return;
		inOrder(node.left);
		System.out.print(node.data + "	");
		inOrder(node.right);
	}

	private void BFStraverse(Node node, List<Node> childNodes) {
		if (node == null)
			return;
		System.out.print(node.data + "	");
		childNodes.add(node.left);
		childNodes.add(node.right);
		while (childNodes.size() > 0) {
			BFStraverse(childNodes.remove(0), childNodes);
		}
	}

	private void printLeafNodes(Node node, List<Node> leafNodes) {
		if (node == null)
			return;
		if (node.left == null && node.right == null) {
			System.out.print(node.data + "	");
		}
		leafNodes.add(node.left);
		leafNodes.add(node.right);
		while (leafNodes.size() > 0) {
			printLeafNodes(leafNodes.remove(0), leafNodes);
		}
	}

	private void printRightNodes(Node node) {
		if (node == null)
			return;
		printRightNodes(node.right);
		if (node.right != null) {
			System.out.print(node.data + "	");
		}
	}

	private void printLeftNodes(Node node) {
		if (node == null)
			return;
		if (node.left != null) {
			System.out.print(node.data + "	");
		}
		printLeftNodes(node.left);
	}
	
	private void BFSWithoutRecursion(Node node) {
		List<Node> nodeList= new ArrayList<>();
		List<Node> childList= new ArrayList<>();
		childList.add(node);
		while(childList.size()>0)
		{
			//retrieve each element
			Node n=childList.remove(0);
			
			//add element's child if not null
			if(n.left!=null)
			{
				childList.add(n.left);
			}
			if(n.right!=null)
			{
				childList.add(n.right);
			}
			// add element into nodeList
			nodeList.add(n);
		}
		
		System.out.println(nodeList);
	}


	public static void main(String[] args) {
		BinarySearchTree obj = new BinarySearchTree();
		obj.insert(5);
		obj.insert(3);
		obj.insert(9);
		obj.insert(1);
		obj.insert(7);
		obj.insert(99);

//		System.out.println("BFS");
//		obj.BFSTraverse();
//		System.out.println();
//		System.out.println("inOrder");
//		obj.inOrder();
//		System.out.println();
//		System.out.println("preOrder");
//		obj.preOrder();
//		System.out.println();
//		System.out.println("postOrder");
//		obj.postOrder();
//		System.out.println();
//		System.out.println("leaf nodes");
//		obj.printLeafNodes();
//		System.out.println();
//		System.out.println("edge nodes");
//		obj.printEdgeNodes();
		obj.BFSWithoutRecursion();

	}
}
